.github/workflows/tox.yml

```diff
@@ -223,9 +223,35 @@ jobs:
         with:
           paths: "**/results.xml"
 
-      - name: Upload to Codecov
+      - name: Upload coverage to artifacts
         # Even if tox fails, upload coverage
         if: ${{ (success() || failure()) && contains(matrix.coverage, 'codecov') && matrix.pytest == 'true' }}
+        uses: actions/upload-artifact@v4
+        with:
+          name: coverage_${{ matrix.toxenv }}.xml
+          path: coverage.xml
+          if-no-files-found: error
+
+  upload-codecov:
+    needs: [tox]
+    permissions:
+      contents: none
+    runs-on: ubuntu-latest
+    name: Upload Coverage
+    steps:
+
+      - name: Download coverage artifacts
+        uses: actions/download-artifact@v4
+        with:
+          path: coverage
+          pattern: coverage_*
+          merge-multiple: true
+
+      - name: Upload report to Codecov
+        if: ${{ hashFiles('coverage/') != '' }}
         uses: codecov/codecov-action@v4
         with:
           token: ${{ secrets.CODECOV_TOKEN }}
+          directory: coverage
```

Comment:
I wonder if we need to specify all the files we downloaded. Unfortunately it seems the download-artifact action only returns the path and not a list of all the artifact files downloaded.

Reply:
Yeah maybe. They have `file` and `files` but the README isn't very clear about usage with wildcards.

Reply:
OK, turns out I misunderstood the settings. The actions won't rename the reports, but I managed to separate them out in different sub-directories. The trick was `merge-multiple: false` and luckily codecov uploader is smart enough to find them anyway.